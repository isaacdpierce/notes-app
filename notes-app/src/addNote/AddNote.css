.add-note {
  background-color: rgb(36, 36, 36);
  position: absolute;
  top: 0;
  left: 0;
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100vh;
  width: 100vw;
  min-width: 330px;
  border: 10px solid rgb(53, 53, 53);
  border-radius: 6px;
}

.add-note__form {
  background-color: var(--main-bg-color);
  text-transform: uppercase;
  display: flex;
  color: rgb(216, 216, 216);
  font-weight: 100;
  justify-content: center;
  align-items: center;
  flex-direction: column;
  padding: 3vw 5vw;
  max-width: 90vw;
  min-width: 310px;
  box-shadow: 0px 1px 1px 1px rgb(17, 17, 17);
}

.add-note__form textarea {
  height: 250px;
}

.add-note__form label {
  display: block;
  margin-bottom: 1rem;
}

.add-note__input {
  font-size: 2rem;
  min-width: 300px;
  width: 50vw;
  background-color: rgb(52, 52, 52);
  color: rgb(216, 216, 216);
  border: none;
  padding: 1.5rem;
}

.add-note__select {
  padding: 2rem 0;
}

.add-note__buttons {
  width: 100%;
  display: flex;
  justify-content: space-between;
}

.add-note__button,
.add-note__link {
  border: 1px solid var(--main-button-color);
  color: var(--main-button-color);
  text-transform: uppercase;
  padding: 0.5rem 1rem;
  font-size: 1.2rem;
  font-weight: 100;
}

.add-note__button {
  background-color: transparent;
}

.folder-select {
  min-width: 150px;
}

.add-note__select {
  text-align: center;
}
