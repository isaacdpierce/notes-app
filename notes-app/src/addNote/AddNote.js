import React, { Component } from 'react';
import { NavLink } from 'react-router-dom';

import NotesContext from '../context/NotesContext';
import ValidationError from '../validationError/ValidationError';

import { config } from '../config/config';

import './AddNote.css';

export default class AddNote extends Component {
  constructor(props) {
    super(props);
    this.state = {
      name: '',
      content: '',
      folderId: '',
      modified: '',
      nameValid: false,
      formValid: false,
      validationMessages: {
        name: '',
      },
    };
  }

  static defaultProps = {
    history: {
      push: () => {},
    },
  };

  static contextType = NotesContext;

  handleNameChange(e) {
    const name = e.target.value;

    this.setState({ name }, () => {
      this.validateName(name);
    });
  }

  handleContentChange(e) {
    const content = e.target.value;
    this.setState({
      content,
    });
  }

  handleSelectChange(e) {
    const folderId = e.target.value;

    this.setState({
      folderId,
    });
  }

  validateName(fieldValue) {
    const fieldErrors = { ...this.state.validationMessages };
    let hasError = false;

    fieldValue = fieldValue.trim();
    if (fieldValue.length === 0) {
      fieldErrors.name = 'Name is required';
      hasError = true;
    } else {
      if (fieldValue.length < 3) {
        fieldErrors.name = 'Name must be at least 3 characters long';
        hasError = true;
      } else {
        fieldErrors.name = '';
        hasError = false;
      }
    }

    this.setState(
      {
        validationMessages: fieldErrors,
        nameValid: !hasError,
      },
      this.formValid
    );
  }

  formValid() {
    this.setState({
      formValid: this.state.nameValid,
    });
  }

  handleAddNoteSubmit(e) {
    const { name, content, folderId } = this.state;

    e.preventDefault();

    const newNote = {
      name,
      content,
      folderId,
      modified: new Date().toISOString(),
    };

    if (!this.state.formValid) {
      if (this.state.validationMessages.name === '')
        this.setState({
          validationMessages: {
            name: 'Please enter a name',
          },
        });
    } else {
      fetch(`${config.API_ENDPOINT}/notes`, {
        method: 'POST',
        headers: {
          'content-type': 'application/json',
        },
        body: JSON.stringify(newNote),
      })
        .then(res => {
          if (!res.ok) return res.json().then(e => Promise.reject(e));
          return res.json();
        })
        .then(note => {
          this.context.addNote(note);
          this.props.history.push(`/folder/${note.folderId}`);
        })
        .catch(error => {
          console.error({ error });
        });
    }
  }

  render() {
    const { folders = [] } = this.context;

    return (
      <section className='add-note'>
        <form action='#' className='add-note__form'>
          <legend>Add a new note</legend>
          <div>
            <label htmlFor='name'>
              name <span className='requiredField'>(required)</span>
            </label>

            <input
              type='text'
              id='name'
              className='add-note__input'
              aria-required='true'
              onChange={e => this.handleNameChange(e)}
            />
            <ValidationError
              hasError={!this.state.nameValid}
              message={this.state.validationMessages.name}
            />
          </div>

          <div>
            <label htmlFor='content'>Content</label>
            <textarea
              type='text'
              id='content'
              className='add-note__input'
              onChange={e => this.handleContentChange(e)}
            />
          </div>
          <div className='add-note__select'>
            <label htmlFor='folder-select'>Select a folder</label>
            <select
              name='folder-select'
              id='folder-select'
              className='folder-select'
              onChange={e => this.handleSelectChange(e)}
            >
              <option value={null}>...</option>
              {folders.map(folder => (
                <option key={folder.id} value={folder.id}>
                  {folder.name}
                </option>
              ))}
            </select>
          </div>
          <div className='add-note__buttons'>
            <NavLink to='/' className='add-note__link'>
              Cancel
            </NavLink>
            <button
              className='add-note__button'
              // disabled={!this.state.formValid}
              onClick={e => this.handleAddNoteSubmit(e)}
            >
              Add note
            </button>
          </div>
        </form>
      </section>
    );
  }
}
