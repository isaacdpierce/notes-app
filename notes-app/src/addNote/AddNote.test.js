import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import AddNote from './AddNote';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

it('renders the complete form', () => {
  const wrapper = shallow(<AddNote />);
  expect(toJson(wrapper)).toMatchSnapshot();
});
