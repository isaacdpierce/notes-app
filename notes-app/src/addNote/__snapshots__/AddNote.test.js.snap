// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the complete form 1`] = `
<section
  className="add-note"
>
  <form
    action="#"
    className="add-note__form"
  >
    <legend>
      Add a new note
    </legend>
    <div>
      <label
        htmlFor="name"
      >
        name 
        <span
          className="requiredField"
        >
          (required)
        </span>
      </label>
      <input
        aria-required="true"
        className="add-note__input"
        id="name"
        onChange={[Function]}
        type="text"
      />
      <ValidationError
        hasError={true}
        message=""
      />
    </div>
    <div>
      <label
        htmlFor="content"
      >
        Content
      </label>
      <textarea
        className="add-note__input"
        id="content"
        onChange={[Function]}
        type="text"
      />
    </div>
    <div
      className="add-note__select"
    >
      <label
        htmlFor="folder-select"
      >
        Select a folder
      </label>
      <select
        className="folder-select"
        id="folder-select"
        name="folder-select"
        onChange={[Function]}
      >
        <option
          value={null}
        >
          ...
        </option>
      </select>
    </div>
    <div
      className="add-note__buttons"
    >
      <NavLink
        className="add-note__link"
        to="/"
      >
        Cancel
      </NavLink>
      <button
        className="add-note__button"
        onClick={[Function]}
      >
        Add note
      </button>
    </div>
  </form>
</section>
`;
